test/document.test.js

```diff
@@ -14907,6 +14951,56 @@ describe('document', function() {
     obj = docNoVersion.toObject();
     assert.ok(!obj.hasOwnProperty('__v'));
   });
+
+  it('allows using overwriteMiddlewareArguments to override pre("init") hook results (gh-15389)', async function() {
+    const timeStringToObject = (time) => {
+      if (typeof time !== 'string') return time;
+      const [hours, minutes] = time.split(':');
+      return { hours: parseInt(hours), minutes: parseInt(minutes) };
+    };
+
+    const timeSchema = new Schema({
+      hours: { type: Number, required: true },
+      minutes: { type: Number, required: true }
+    });
+
+    // Attempt to transform during init
+    timeSchema.pre('init', function(rawDoc) {
+      if (typeof rawDoc === 'string') {
+        return mongoose.overwriteMiddlewareArguments(timeStringToObject(rawDoc));
+      }
+    });
+
+    const userSchema = new Schema({
+      unknownKey: {
+        type: timeSchema,
+        required: true
+      }
+    });
+    const User = db.model('Test', userSchema);
+    await User.collection.insertOne({ unknownKey: '12:34' });
+    const user = await User.findOne();
```

Comment:
Just to ensure no flaky tests (residual data) happen, maybe consider finding by id instead.